crates/ruff_python_parser/src/lexer.rs

```diff
@@ -428,6 +429,40 @@ impl<'source> Lexer<'source> {
                     self.cursor.bump();
                     value.push('\\');
                 }
+                '?' => {
+                    self.cursor.bump();
+                    let mut question_count: u8 = 1;
+                    while self.cursor.eat_char('?') {
+                        question_count += 1;
+                    }
+
+                    // The help end magic command is valid if it has 1 or 2 question marks.
+                    // If the `value` is empty i.e., we are at the start of the line,
+                    // then we can't be sure if it is a help end magic command or not.
```

Comment:
It took me a very long time to understand this part (I even had to make code changes). It may be helpful to add an inline comment with a few code examples of what we're solving here.

Reply:
Oh sorry that this was unclear. I'll add some additional comments, references and examples to make it understandable.